components/calendar/calendar.js

```diff
@@ -0,0 +1,488 @@
+
+	if (numDaysFromPrevMonthToShow > 0) {
+		const prevMonth = getPrevMonth(shownMonth);
+		const prevMonthYear = (shownMonth === 0) ? (shownYear - 1) : shownYear;
+		const numDaysLastMonth = getNumberOfDaysInMonth(prevMonth, prevMonthYear);
+		for (let i = numDaysLastMonth - numDaysFromPrevMonthToShow + 1; i <= numDaysLastMonth; i++) {
+			firstWeek.push(new Date(prevMonthYear, prevMonth, i));
+		}
+	}
+	for (let j = 1; j <= daysInWeek - numDaysFromPrevMonthToShow; j++) {
+		firstWeek.push(new Date(shownYear, shownMonth, j));
+	}
+	dates.push(firstWeek);
+
+	// remaining weeks
+	let nextMonthDay = 1;
+	let firstDateOfWeek = daysInWeek - numDaysFromPrevMonthToShow + 1;
+	const numWeeks = Math.ceil((numDaysFromPrevMonthToShow + numDays) / daysInWeek);
+	const nextMonth =  getNextMonth(shownMonth);
+	const nextMonthYear = (shownMonth === 11) ? (shownYear + 1) : shownYear;
+	for (let i = 1; i < numWeeks; i++) {
+		const week = [];
+		for (let j = firstDateOfWeek; j < firstDateOfWeek + daysInWeek; j++) {
+			let day;
+			if (j > numDays) {
+				day = new Date(nextMonthYear, nextMonth, nextMonthDay);
+				nextMonthDay++;
+			} else {
+				day = new Date(shownYear, shownMonth, j);
+			}
+			week.push(day);
+		}
+		firstDateOfWeek = firstDateOfWeek + daysInWeek;
+		dates.push(week);
+	}
+
+	return dates;
+}
+
+export function getNextMonth(month) {
+	return (month === 11) ? 0 : (month + 1);
+}
+
+export function getNumberOfDaysFromPrevMonthToShow(month, year) {
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	let numDaysFromLastMonthToShowThisMonth = 0;
+	if (firstDayOfMonth !== firstDayOfWeek) {
+		numDaysFromLastMonthToShowThisMonth = firstDayOfMonth - firstDayOfWeek;
+		if (numDaysFromLastMonthToShowThisMonth < 0) {
+			numDaysFromLastMonthToShowThisMonth += daysInWeek;
+		}
+	}
+	return numDaysFromLastMonthToShowThisMonth;
+}
+
+export function getNumberOfDaysInMonth(month, year) {
+	return new Date(year, month + 1, 0).getDate();
+}
+
+export function getNumberOfDaysToSameWeekPrevMonth(month, year) {
+	const prevMonth = getPrevMonth(month);
+	const prevMonthYear = prevMonth === 11 ? year - 1 : year;
+	const numDaysPrevMonth = getNumberOfDaysInMonth(prevMonth, prevMonthYear);
+	const numDaysFromPrevPrevMonthToShowPrevMonth = getNumberOfDaysFromPrevMonthToShow(prevMonth, prevMonthYear);
+	const numWeeksPrevMonth = Math.ceil((numDaysFromPrevPrevMonthToShowPrevMonth + numDaysPrevMonth) / 7);
+
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	const hasDaysFromPrevMonth = (firstDayOfMonth !== firstDayOfWeek);
+	return daysInWeek * (numWeeksPrevMonth - ((hasDaysFromPrevMonth) ? 1 : 0));
+}
+
+export function getPrevMonth(month) {
+	return (month === 0) ? 11 : (month - 1);
+}
+
+export function getToday() {
+	const val = new Date().toISOString();
+	const re = /([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{1,2}):([0-9]{1,2}):([0-9]{1,2})/;
+	const match = val.match(re);
```

Comment:
Why do you need the regular expression here to extra the different date parts? Couldn't you just use the `Date` methods? Oh I guess it's because it'll be in local time but `toISOString` converts it to UTC?

Reply:
Yeah exactly